lib/os/queued_operation.c

```diff
@@ -0,0 +1,216 @@
+		 * completes finalization, lest we pick something that
+		 * has too low a priority.
+		 */
+		if (mgr->finalizing) {
+			break;
+		}
+
+		sys_snode_t *node = sys_slist_get(&mgr->operations);
+
+		if (node) {
+			op = CONTAINER_OF(node, struct queued_operation, node);
+			mgr->current = op;
+		} else {
+			op = NULL;
+		}
+
+		k_spin_unlock(&mgr->lock, key);
+
+		/* Only notify the manager if there's an operation, or
+		 * if it is to transition to idle.
+		 */
+		if ((op != NULL) || !in_idle) {
+			mgr->vtable->process(mgr, op);
+		}
+		in_idle = (op == NULL);
+
+		key = k_spin_lock(&mgr->lock);
+	} while (op != NULL);
+
+	k_spin_unlock(&mgr->lock, key);
+}
+
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority)
+{
+	int rv = 0;
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(mgr->vtable != NULL);
+	__ASSERT_NO_MSG(mgr->vtable->process != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	/* Validation is optional; if present, use it. */
+	if (mgr->vtable->validate) {
+		rv = mgr->vtable->validate(mgr, op);
+	}
+
+	/* Set the priority, checking whether it's in range. */
+	if (rv == 0) {
+		rv = op_set_priority(op, priority);
+	}
+
+	/* Reject callback notifications without translation
+	 * function.
+	 */
+	if ((rv == 0)
+	    && async_notify_uses_callback(&op->notify)
+	    && (mgr->vtable->callback == NULL)) {
+		rv = -ENOTSUP;
+	}
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	sys_slist_t *list = &mgr->operations;
+
+	if (priority == QUEUED_OPERATION_PRIORITY_PREPEND) {
+		sys_slist_prepend(list, &op->node);
+	} else if (priority == QUEUED_OPERATION_PRIORITY_APPEND) {
+		sys_slist_append(list, &op->node);
+	} else {
+		struct queued_operation *prev = NULL;
+		struct queued_operation *tmp;
+
+		SYS_SLIST_FOR_EACH_CONTAINER(list, tmp, node) {
+			if (priority < op_get_priority(tmp)) {
+				break;
+			}
+			prev = tmp;
+		}
+
+		if (prev == NULL) {
+			sys_slist_prepend(list, &op->node);
+		} else {
+			sys_slist_insert(list, &prev->node, &op->node);
+		}
+	}
+
+	select_next_and_unlock(mgr, key);
+
+	return rv;
+}
+
+static inline void finalize_and_notify(struct queued_operation_manager *mgr,
+				       struct queued_operation *op,
+				       int res)
+{
+	async_notify_generic_callback cb = async_notify_finalize(&op->notify, res);
+
+	if (cb != NULL) {
+		mgr->vtable->callback(mgr, op, cb);
+	}
+}
+
+void queued_operation_finalize(struct queued_operation_manager *mgr,
+			       struct queued_operation *op,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	if (mgr->current == op) {
+		mgr->finalizing = true;
+		mgr->current = NULL;
+	}
+
+	k_spin_unlock(&mgr->lock, key);
+
+	finalize_and_notify(mgr, op, res);
+
+	key = k_spin_lock(&mgr->lock);
+
+	mgr->finalizing = false;
+
+	if (can_start(mgr)) {
+		select_next_and_unlock(mgr, key);
```

Comment:
At this point it is already known that `mgr->current` is `NULL` and `mgr->finalizing` is `false`, but both `can_start` and `select_next_and_unlock` will check those (the checks from `can_start` will be probably optimized by the compiler).
I'd propose to refactor this a bit and do these checks only in `queued_operation_submit`, right before calling `select_next_and_unlock`, and here to call `select_next_and_unlock` directly.

Reply:
I've just removed the extra check and called `select_next_and_unlock()` in both places.  I don't think there's any other refactoring that would be safe: `select_next_and_unlock()` needs the check to behave properly with operations that complete during the initial process call.